larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -58,12 +68,81 @@ DenseElementsAttr Bitpack(PatternRewriter& builder, Attribute x) {
 
 #include "larq_compute_engine/mlir/transforms/generated_optimize.inc"
 
+struct SetBconvReadWriteBitpacked : public OpRewritePattern<TF::LceBconv2dOp> {
+  using OpRewritePattern<TF::LceBconv2dOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(TF::LceBconv2dOp bconv_op,
+                                     PatternRewriter& rewriter) const override {
+    Value bconv_input = bconv_op.input();
+    if (!bconv_input.hasOneUse()) return matchFailure();
+
+    auto bconv_input_op =
+        dyn_cast_or_null<TF::LceBconv2dOp>(bconv_input.getDefiningOp());
+    if (!bconv_input_op) return matchFailure();
+
+    const auto inner_tensor_type = bconv_input_op.getType().cast<ShapedType>();
+
+    // We use 32-bit bitpacking.
+    constexpr int bitwidth = 32;
+
+    // We can only apply this transformation if the inner tensor type is F32.
+    if (inner_tensor_type.getElementType().isInteger(bitwidth))
```

Comment:
Do we need to check if `bconv_op.channels_in()` is set, or will the previous pattern already catch that?

Reply:
I don't think this is an issue if we're requiring the attribute to be set?

Reply:
Yeah I think you're right. I was checking the pattern and it will always be set in the prepare pass.